### PR TITLE
[16.0][IMP]l10n_it_delivery_note: create invoice from delivery note

### LINKED ISSUE #3979: [16.0]l10n_it_delivery_note: creando una fattura da più ddt le righe non vengono generate basandosi sulle righe dei ddt
In riferimento alla pr: https://github.com/OCA/l10n-italy/pull/3552
Volendo creare una fattura da più ddt, al momento, la fattura viene generata tenendo conto delle righe dell'ordine di vendita.
In questo modo nelle fatture e nella stampa non vengono generate più righe basandosi sul ddt
Sarebbe utile generare invece una fattura con più righe per ogni riga nei vari ddt:
image

### PR DESCRIPTION
**Pull request by odooNextev**
Al momento non è possibile creare una fattura direttamente da ddt come nelle precedenti versione. (#3552 )
Creazione fattura con più righe per ogni riga nei vari ddt:(#3979 )

### CODE REVIEW

**Review by MaurizioPellegrinet — APPROVED** [review]
LGTM

**Comment by francesco-ooops** [comment]
@odooNextev questa PR è linkata ad una PR linkata ad un'altra PR, riusciresti ad aprire una nuova issue per spiegare in breve cosa andiamo a risolvere con questa?

Grazie mille :)

**Comment by TheMule71** [comment]
Vd #4188

**Comment by odooNextev** [comment]
@francesco-ooops ora c'è la issue

**Comment by odooNextev** [comment]
@francesco-ooops @Borruso Ora i test passano. Riuscireste a fare una review? Grazie mille

**Comment by francesco-ooops** [comment]
@odooNextev scusa mi son perso nella catena di issues e PR, mi daresti due indicazioni rapide per testare funzionalmente?

**Comment by odooNextev** [comment]
Per testare dovresti generare una fattura da più ddt, dovresti ottenere le righe raggruppate per ogni delivery note. Nei settings trovi una checkbox dedicata "Group Delivery note invoices by quantity" che viene utilizzata per gestire questo caso oppure mantenere il metodo attuale.

**Comment by francesco-ooops** [comment]
@odooNextev puoi aggiungere il funzionamento alla documentazione?

File: l10n_it_delivery_note/models/res_config_settings.py [hunk 1]
```diff
@@ -49,3 +49,6 @@ def _default_virtual_locations_root(self):
         related="company_id.display_delivery_method_dn_report",
         readonly=False,
     )
+    delivery_note_group_by_quantity = fields.Boolean(
+        string="Group Delivery note invoices by quantity"
```
Comment (Borruso) [inline l10n_it_delivery_note/models/res_config_settings.py]: dovresti mettere il related con la company come fatto per gli altri campi

**Review by Borruso — CHANGES_REQUESTED** [review]
Grazie mille per la PR
Molto ìnteressante

File: l10n_it_delivery_note/models/res_config_settings.py (hunk 1 above)
Comment (odooNextev) [reply l10n_it_delivery_note/models/res_config_settings.py]: Fatto

**Review by Borruso — APPROVED** [review]
LGTM

**Review by francesco-ooops — CHANGES_REQUESTED** [review]
@odooNextev puoi aggiungere il funzionamento alla documentazione?

**Comment by odooNextev** [comment]
> @odooNextev puoi aggiungere il funzionamento alla documentazione?

è sufficiente?

**Review by MarcoCalcagni — APPROVED** [review]
OK in uso presso cliente

**Comment by MarcoCalcagni** [comment]
> @odooNextev puoi aggiungere il funzionamento alla documentazione?

puoi aggiornare l'Approve ?

**Comment by francesco-ooops** [comment]
La documentazione è ok, al momento non riesco a fare test funzionale

**Review by MaurizioPellegrinet — APPROVED** [review]
Funzionale OK

**Review by francesco-ooops — APPROVED** [review]
A parte la documentazione con un file in italiano e uno in inglese ma vabè 😅 (non dipende certo da voi)

Funzionalmente mi sembra lineare anche se potrei essermi perso qualche casistica, mergerei 

in ogni caso credo che il funzionamento senza il flag attivato sia da fixare:

1 ordine con 1 riga > 2 DDT > 1 fattura al momento risulta così nel runboat 16 (mentre nella 14 c'è il bug segnalato in #4309 )

image

**Review by francesco-ooops — CHANGES_REQUESTED** [review]
@odooNextev come dicevo funzionale ok, tuttavia a quasi quattro anni di distanza dal merge di questo modulo stiamo ancora lottando per avere il riferimento corretto del DDT in fattura (decisamente non una funzionalità secondaria e vari clienti ce l'hanno fatto notare con fastidio). 

Mi sembra evidente che su questo tema non ci sia stata una analisi delle casistiche da gestire per poter aggiungere dei test che evitino regressioni, sarebbe cosa molto utile aggiungerle in questa PR, che ne dite?

Nel mentre stiamo facendo lo stesso per assicurare che funzioni il riferimento DDT in fattura anche senza questo flag attivo, appena fatto riportiamo tutto nella v16 👍

**Comment by odooNextev** [comment]
@francesco-ooops ok allora aggiungo dei test per la parte che abbiamo introdotto

**Comment by francesco-ooops** [comment]
@odooNextev in realtà ho riscontrato una modifica al funzionamento standard in questa PR, passi per riprodurre: 

- Disabilito flag "Group Delivery note invoices by quantity" (tra l'altro lo rinominerei "Group invoice lines by delivery note")

SO, product 1, qty 2
Consegno qty 1, creo backorder, valido DDT1
Consegno qty 1, valido DDT2
Fatturo da SO

Prima della PR: 

image

Dopo la PR

image

**Comment by odooNextev** [comment]
@francesco-ooops ho aggiunto un test abbastanza corposo.

> flag "Group Delivery note invoices by quantity" (tra l'altro lo rinominerei "Group invoice lines by delivery note")

Stai testando nel runboat?
Già dal push precedente avevo modificato il nome del parametro.

Hai ragione sullo spostamento delle righe delle note, devo sistemare.

**Review by francesco-ooops — CHANGES_REQUESTED** [review]
@odooNextev steps to reproduce: 

Abilito flag "Group invoice lines by Delivery Note"
Creo SO001 con prodotto 1, qty 2, valido
Spedisco prodotto 1, qty 1 (creo backorder) > creo e valido DDT001 > creo fattura INV001 da ordine
Spedisco prodotto 1, qty 1 > creo e valido DDT002 > creo fattura INV002 da ordine

Current behavior

La seconda fattura INV002 non ha riferimento al DDT ma ha due volte la riga prodotto (?)

Required behavior

INV001 ha riferimento a DDT001, INV002 ha riferimento a DDT002

@MarcoCalcagni non so se è così sicuro tenerlo in ambiente di produzione al momento

**Comment by francesco-ooops** [comment]
@stenext pingami quando serve che testo :)

**Comment by odooNextev** [comment]
> @stenext pingami quando serve che testo :)

Sì, devo correggere un paio di cose perciò ti avviso quando ho fatto, non fare affidamento sui push

File: l10n_it_delivery_note/models/sale_order.py [hunk 2]
```diff
@@ -121,13 +124,138 @@ def _generate_delivery_note_lines(self, invoice_ids):
...
+        for move in moves:
+            move.message_post_with_view(
+                "mail.message_origin_link",
+                values={"self": move, "origin": move.line_ids.sale_line_ids.order_id},
+                subtype_id=self.env["ir.model.data"]._xmlid_to_res_id("mail.mt_note"),
+            )
+        # return moves
+        self._assign_delivery_notes_invoices(moves.ids)
+        self._generate_delivery_note_lines(moves.ids)
+        return moves
```
Comment (HekkiMelody) [inline l10n_it_delivery_note/models/sale_order.py]: **issue**: questa implementazione (che riprende, con piccole modifiche, il metodo per come scritto nel modulo `sale`) non chiama mai `super()`, rompendo tutti gli altri moduli che personalizzano `_create_invoices()`, tra cui i moduli core `sale_timesheet` e `repair` e svariati moduli OCA

**Review by HekkiMelody — CHANGES_REQUESTED** [review]

File: l10n_it_delivery_note/models/sale_order.py (hunk 2 above)
Comment (odooNextev) [reply l10n_it_delivery_note/models/sale_order.py]: Potrebbe andare bene una soluzione del genere?
Chiamare la `super()` nel caso in cui ci sia attivata l'opzione è quasi impossibile perchè andrebbe fatta a priori per poi distruggere tutto quello che ha fatto per sostituire le righe con il nuovo raggruppamento.

L'unica cosa è che diventerebbe impossibile richiamare il metodo originale nel caso il flag non fosse attivo ed ho dovuto riportare il metodo originale e non mi piace molto...

Comment (HekkiMelody) [reply l10n_it_delivery_note/models/sale_order.py]: Con un monkeypatching di questo tipo, credo non bisognerebbe fare un nuovo model ma solo definire la funzione (fuori dalla classe), e poi andarla a sostituire alla funzione originale come hai fatto in fondo.

A mali estremi, potrebbe funzionare. Di certo la funzione originale è troppo lunga per andarsi a inserire in maniera efficace.

Abbiamo provato un'implementazione alternativa per la stessa idea di fondo (1 riga DDT = 1 riga fattura) che speriamo essere più compatibile qui https://github.com/OCA/l10n-italy/pull/4324 (Non è ancora pronta, ci sono test da aggiungere e edge cases da limare)
